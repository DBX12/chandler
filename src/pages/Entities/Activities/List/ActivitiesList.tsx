import moment from 'moment';
import React, {PureComponent} from 'react';
import {ActivityIndicator, FlatList, Text, View} from 'react-native';

import {EmptyActivity, LastTwoYearsStatistics, Navbar} from '@components';
import {I18n} from '@i18n';
import {IActivity, IContact, IMetaStatistics} from '@models';
import {IPopAction} from '@navigator/NavigationService';
import {commonStyles} from '@theme';
import {styles} from './ActivitiesList.styles';

interface IActivitiesListProps {
  pop: IPopAction; // TODO: fix this
  getActivitiesByContact: () => void;
  contact: IContact;
  activities: IActivity[];
  isFetching: boolean;
  statistics: IMetaStatistics;
}

export class ActivitiesList extends PureComponent<IActivitiesListProps, {}> {
  public componentWillMount() {
    this.props.getActivitiesByContact();
  }

  public keyExtractor = (item, index) => String(item.id);

  public renderHeader = () => {
    const {isFetching, statistics} = this.props;

    if (isFetching) {
      return null;
    }

    const keys = Object.keys(statistics).reverse();

    return (
      <View style={styles.headerContainer}>
        <LastTwoYearsStatistics
          image={require('@assets/icons/activities.png')}
          title1={I18n.t('common:yearDisplay', {year: keys[0]})}
          count1={statistics[keys[0]]}
          title2={I18n.t('common:yearDisplay', {year: keys[1]})}
          count2={statistics[keys[1]]}
        />
      </View>
    );
  };

  public renderFooter = () => {
    const {isFetching} = this.props;

    if (!isFetching) {
      return null;
    }

    return (
      <ActivityIndicator style={commonStyles.activityIndicator} size="large" />
    );
  };

  public renderItem = ({item, index}) => {
    const {activities} = this.props;
    const activity = activities[index];

    return (
      <View style={styles.activityContainer}>
        <View style={commonStyles.row}>
          <Text style={styles.textLeft}>{activity.summary}</Text>
          <View style={commonStyles.flex} />
          <Text style={styles.textRight}>
            {moment(activity.date_it_happened).fromNow()}
          </Text>
        </View>
        {activity.description && (
          <Text style={styles.textInfo}>{activity.description}</Text>
        )}
      </View>
    );
  };

  public render() {
    const {
      pop,
      activities,
      getActivitiesByContact,
      isFetching,
      contact,
    } = this.props;

    return (
      <View style={commonStyles.flex}>
        <Navbar title={I18n.t('activities:activities')} onBack={pop} />
        {isFetching || activities.length ? (
          <FlatList
            data={activities}
            renderItem={this.renderItem}
            keyExtractor={this.keyExtractor}
            ListHeaderComponent={this.renderHeader}
            ListFooterComponent={this.renderFooter}
            onEndReached={getActivitiesByContact}
            onEndReachedThreshold={0.5}
          />
        ) : (
          <EmptyActivity
            image={require('./assets/empty-activities.png')}
            title={I18n.t('activities:emptyTitle', {name: contact.first_name})}
            subtitle={I18n.t('activities:emptySubtitle')}
          />
        )}
      </View>
    );
  }
}
