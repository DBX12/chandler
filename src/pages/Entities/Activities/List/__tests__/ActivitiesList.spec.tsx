import {shallow} from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import {ActivitiesList} from '../ActivitiesList';

describe('Pages', () => {
  describe('ActivitiesList', () => {
    let pop;
    let getActivitiesByContact;
    const contact = {first_name: 'Theo'};

    beforeEach(() => {
      pop = jest.fn();
      getActivitiesByContact = jest.fn();
    });

    it('should renders correctly without activities and fetching', () => {
      const activities = [];
      const statistics = {'2018': 100, '2015': 1};
      const tree = shallow(
        <ActivitiesList
          contact={contact as any}
          pop={pop}
          getActivitiesByContact={getActivitiesByContact}
          isFetching={true}
          activities={activities}
          statistics={statistics}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
    });

    it('should renders correctly without activities and not fetching', () => {
      const activities = [];
      const statistics = {};
      const tree = shallow(
        <ActivitiesList
          pop={pop}
          contact={contact as any}
          getActivitiesByContact={getActivitiesByContact}
          isFetching={false}
          activities={activities}
          statistics={statistics}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
    });

    it('should renders correctly with activities and fetching', () => {
      const activities = [
        {
          summary: 'Activity title',
          description: 'My description',
          date_it_happened: '1991-12-17',
        },
      ];
      const statistics = {'2018': 100, '2015': 1};
      const tree = shallow(
        <ActivitiesList
          contact={contact as any}
          pop={pop}
          getActivitiesByContact={getActivitiesByContact}
          isFetching={true}
          activities={activities as any}
          statistics={statistics}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 0})))
      ).toMatchSnapshot();
    });

    it('should renders correctly with activities and fetching', () => {
      const activities = [
        {
          summary: 'Activity title',
          description: 'My description',
          date_it_happened: '1991-12-17',
        },
        {
          summary: 'Activity title 2',
          description: 'My description 2',
          date_it_happened: '1993-12-17',
        },
      ];
      const statistics = {'2018': 100, '2015': 1};
      const tree = shallow(
        <ActivitiesList
          contact={contact as any}
          pop={pop}
          getActivitiesByContact={getActivitiesByContact}
          isFetching={false}
          activities={activities as any}
          statistics={statistics}
        />
      );
      expect(toJson(tree)).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 0})))
      ).toMatchSnapshot();
      expect(
        toJson(shallow(tree.instance().renderItem({index: 1})))
      ).toMatchSnapshot();
    });

    it('should not display the header if it is fetching', () => {
      const activities = [
        {
          summary: 'Activity title',
          description: 'My description',
          date_it_happened: '1991-12-17',
        },
        {
          summary: 'Activity title 2',
          description: 'My description 2',
          date_it_happened: '1993-12-17',
        },
      ];
      const statistics = {'2018': 100, '2015': 1};
      const tree = shallow(
        <ActivitiesList
          contact={contact as any}
          pop={pop}
          getActivitiesByContact={getActivitiesByContact}
          isFetching={true}
          activities={activities as any}
          statistics={statistics}
        />
      );
      expect(tree.instance().renderHeader()).toBeNull();
    });

    it('should display the header if it is not fetching', () => {
      const activities = [
        {
          summary: 'Activity title',
          description: 'My description',
          date_it_happened: '1991-12-17',
        },
        {
          summary: 'Activity title 2',
          description: 'My description 2',
          date_it_happened: '1993-12-17',
        },
      ];
      const statistics = {'2017': 100, '2015': 20};
      const tree = shallow(
        <ActivitiesList
          contact={contact as any}
          pop={pop}
          getActivitiesByContact={getActivitiesByContact}
          isFetching={false}
          activities={activities as any}
          statistics={statistics}
        />
      );
      expect(toJson(shallow(tree.instance().renderHeader()))).toMatchSnapshot();
    });

    it('should display the activity indicator in the footer if it is fetching', () => {
      const activities = [
        {
          summary: 'Activity title',
          description: 'My description',
          date_it_happened: '1991-12-17',
        },
        {
          summary: 'Activity title 2',
          description: 'My description 2',
          date_it_happened: '1993-12-17',
        },
      ];
      const statistics = {'2017': 2300, '2015': 10};
      const tree = shallow(
        <ActivitiesList
          contact={contact as any}
          pop={pop}
          getActivitiesByContact={getActivitiesByContact}
          isFetching={true}
          activities={activities as any}
          statistics={statistics}
        />
      );
      expect(toJson(shallow(tree.instance().renderFooter()))).toMatchSnapshot();
    });

    it('should not display the activity indicator in the footer if it is not fetching', () => {
      const activities = [
        {
          summary: 'Activity title',
          description: 'My description',
          date_it_happened: '1991-12-17',
        },
        {
          summary: 'Activity title 2',
          description: 'My description 2',
          date_it_happened: '1993-12-17',
        },
      ];
      const statistics = {'2020': 100, '2015': 20};
      const tree = shallow(
        <ActivitiesList
          contact={contact as any}
          pop={pop}
          getActivitiesByContact={getActivitiesByContact}
          isFetching={false}
          activities={activities as any}
          statistics={statistics}
        />
      );
      expect(tree.instance().renderFooter()).toBeNull();
    });

    it('should have a keyExtractor', () => {
      const activities = [{id: 1}, {id: 2}];
      const statistics = {'2017': 100, '2015': 20};
      const tree = shallow(
        <ActivitiesList
          contact={contact as any}
          pop={pop}
          getActivitiesByContact={getActivitiesByContact}
          isFetching={false}
          activities={activities as any}
          statistics={statistics}
        />
      );
      expect(tree.instance().keyExtractor(activities[0])).toBe('1');
      expect(tree.instance().keyExtractor(activities[1])).toBe('2');
    });
  });
});
