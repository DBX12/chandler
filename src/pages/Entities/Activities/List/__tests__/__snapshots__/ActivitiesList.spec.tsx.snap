// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pages ActivitiesList should display the activity indicator in the footer if it is fetching 1`] = `
<ActivityIndicator
  animating={true}
  color="#999999"
  hidesWhenStopped={true}
  size="large"
  style={
    Object {
      "marginVertical": 20,
    }
  }
/>
`;

exports[`Pages ActivitiesList should display the header if it is not fetching 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "paddingHorizontal": 15,
    }
  }
>
  <LastTwoYearsStatistics
    count1={100}
    count2={20}
    image={1}
    title1="In 2017"
    title2="In 2015"
  />
</View>
`;

exports[`Pages ActivitiesList should renders correctly with activities and fetching 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Activities"
  />
  <FlatList
    ListFooterComponent={[Function]}
    ListHeaderComponent={[Function]}
    data={
      Array [
        Object {
          "date_it_happened": "1991-12-17",
          "description": "My description",
          "summary": "Activity title",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</Component>
`;

exports[`Pages ActivitiesList should renders correctly with activities and fetching 2`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "paddingHorizontal": 15,
      "paddingVertical": 20,
    }
  }
>
  <Component
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#355974",
        }
      }
    >
      Activity title
    </Text>
    <Component
      style={
        Object {
          "flex": 1,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#97A5B2",
        }
      }
    >
      25 years ago
    </Text>
  </Component>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 14,
        "marginTop": 5,
      }
    }
  >
    My description
  </Text>
</View>
`;

exports[`Pages ActivitiesList should renders correctly with activities and fetching 3`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Activities"
  />
  <FlatList
    ListFooterComponent={[Function]}
    ListHeaderComponent={[Function]}
    data={
      Array [
        Object {
          "date_it_happened": "1991-12-17",
          "description": "My description",
          "summary": "Activity title",
        },
        Object {
          "date_it_happened": "1993-12-17",
          "description": "My description 2",
          "summary": "Activity title 2",
        },
      ]
    }
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</Component>
`;

exports[`Pages ActivitiesList should renders correctly with activities and fetching 4`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "paddingHorizontal": 15,
      "paddingVertical": 20,
    }
  }
>
  <Component
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#355974",
        }
      }
    >
      Activity title
    </Text>
    <Component
      style={
        Object {
          "flex": 1,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#97A5B2",
        }
      }
    >
      25 years ago
    </Text>
  </Component>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 14,
        "marginTop": 5,
      }
    }
  >
    My description
  </Text>
</View>
`;

exports[`Pages ActivitiesList should renders correctly with activities and fetching 5`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "borderBottomWidth": 0.5,
      "borderColor": "#E2E4E5",
      "paddingHorizontal": 15,
      "paddingVertical": 20,
    }
  }
>
  <Component
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#355974",
        }
      }
    >
      Activity title 2
    </Text>
    <Component
      style={
        Object {
          "flex": 1,
        }
      }
    />
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "#97A5B2",
        }
      }
    >
      23 years ago
    </Text>
  </Component>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    style={
      Object {
        "fontSize": 14,
        "marginTop": 5,
      }
    }
  >
    My description 2
  </Text>
</View>
`;

exports[`Pages ActivitiesList should renders correctly without activities and fetching 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Activities"
  />
  <FlatList
    ListFooterComponent={[Function]}
    ListHeaderComponent={[Function]}
    data={Array []}
    disableVirtualization={false}
    horizontal={false}
    initialNumToRender={10}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onEndReached={
      [MockFunction] {
        "calls": Array [
          Array [],
        ],
        "results": Array [
          Object {
            "isThrow": false,
            "value": undefined,
          },
        ],
      }
    }
    onEndReachedThreshold={0.5}
    renderItem={[Function]}
    scrollEventThrottle={50}
    updateCellsBatchingPeriod={50}
    windowSize={21}
  />
</Component>
`;

exports[`Pages ActivitiesList should renders correctly without activities and not fetching 1`] = `
<Component
  style={
    Object {
      "flex": 1,
    }
  }
>
  <Navbar
    onBack={[MockFunction]}
    title="Activities"
  />
  <EmptyActivity
    image={1}
    subtitle="Keep track of what youâ€™ve done."
    title="Have you done an activity with Theo lately?"
  />
</Component>
`;
